use iso8601 duration info from videoData to display video duration in parens after the title

jmartin-sul opened the pull request:
this adds a new [dependency](https://github.com/arnau/ISO8601) because:
* ruby's time libraries don't parse ISO 8601 durations
* ActiveSupport's Duration class does, but not until rails 5
* arnau's library seems well tested, and a better fit than the other options i looked at:
  * https://github.com/peleteiro/ruby-duration
  * https://github.com/janko-m/as-duration (just reimplements that AS4 duration)

AFAICT, nothing off the shelf provides for nice human readable printing of a duration object in the way you'd expect a media timer to do it.  as the mockup implies, this will print durations such that:
* all fields after the first non-zero field are included
* all fields after the first included field are zero padded

e.g.:
`P0DT1H2M3S` => `1:02:03` (i.e. 0 days, 1 hour, 2 minutes, 3 seconds)
`PT10S` => `0:10` (i.e. 10 seconds)
`P9D` => `9:00:00:00` (i.e. 9 days)

before:
<img width="1439" alt="screen shot 2016-08-11 at 2 32 44 pm" src="https://cloud.githubusercontent.com/assets/7741604/17605713/a6e2def2-5fd0-11e6-9132-5f27d4c4d2e1.png">

after (title on last object pushes duration out of view; @jkeck and i were inclined to say we can live with that, but also wanted input from @jvine to see if there was a better solution or a placement that'd keep that from happening):
<img width="1450" alt="screen shot 2016-08-11 at 2 33 24 pm" src="https://cloud.githubusercontent.com/assets/7741604/17605720/ada32cf6-5fd0-11e6-98db-3d4c9d9ae31b.png">



see also: https://en.wikipedia.org/wiki/ISO_8601#Durations (it turns out the actual ISO doc costs 138 swiss francs)

closes issue #636 
closes issue #574 
connect to sul-dlss/sul-embed#636 
connect to sul-dlss/sul-embed#574

jmartin-sul commented:
attempted to break up into a reasonable set of logical commits.  can squash further if people see reason to do so.

jkeck commented on app/models/embed/purl.rb:
```diff
@@ -137,6 +138,30 @@ def initialize(msg = 'The requested PURL resource was not embeddable.')
+    class Duration < ISO8601::Duration
+      def to_s
+        raise 'Duration#to_s does not support specifying durations in weeks' if atoms[:weeks] != 0
+        raise 'Duration#to_s does not support specifying negative durations' if atoms.values.detect { |val| val < 0 }
```
Do we really want to raise an error in this case?  That would, AFAICT, cause the embed viewer to error out and not display anything for these objects.  Should we consider a more graceful approach for when the data is not what we want/expect?

jkeck replied:
If we do keep these checks, I think `atoms.values.any? { |val| val < 0 }` might be more idiomatic than `atoms.values.detect { |val| val < 0 }`.

jmartin-sul replied:
ah, yeah, fair point about the erroring out of the whole viewer, hadn't thought about that.

would you keep the check and do something milder, like logging?  or just ditch it entirely?

and yeah, i like `any?`, were we to keep the check, though i'm happy to get rid of it.

jmartin-sul replied:
i guess another option would be some sort of text indicator of the error, like returning `--`, though that seems like it might just confuse users.

jkeck replied:
Yeah, I think logging something.  I'm not sure if there is a way in honeybadger (which we aren't even setup for yet, but no sense in looking into squash) to tell it to log/track an issue that doesn't necessarily raise an error through the actual stack.

But in the meantime, I think logging it would be good and just printing nothing for the duration.

jvine commented:
> title on last object pushes duration out of view

Hard to do gracefully without some kind of calculation to add ellipsis to the title. Pin the duration to the 4th row, so it always appears in the same place; and if the title is longer than 3 rows (or 2 if it's restricted) it would be truncated. 

<img width="534" alt="ellipsis" src="https://cloud.githubusercontent.com/assets/6945691/17606463/855f2980-5fd4-11e6-8284-c7a1f4abfade.png">

(And maybe expose the full title on hover/focus...?)

jmartin-sul commented:
ok, updated to address feedback.  squashed fixes for @jkeck's suggestions into existing commits.

added a new commit for something like what @jvine suggested, but instead of limiting the text to 3 lines, it truncates it with an ellipsis at 45 chars.  this seems to display well in my manual testing, and as seen in the screenshot, there are at least some examples where the implemented solution shows more of the title than would otherwise be shown. but if that solution is unacceptable, it's easy enough to remove the commit.

<img width="1442" alt="screen shot 2016-08-11 at 7 58 47 pm" src="https://cloud.githubusercontent.com/assets/7741604/17611736/b7dfb666-5ffe-11e6-9a5d-15c4d94774ee.png">

i didn't implement the full text on hover/focus suggestion, but would be happy to look into it (though maybe we should do so under a separate ticket?).

jmartin-sul commented:
also:  github is showing commits ba705f0 and 0569a11 in the wrong order:  the refactor in ba705f0 comes first, and should show up that way if you check out the branch and look at the log on the command line (ba705f0 still has an earlier timestamp, which i imagine is the cause of github's display confusion).

jkeck commented on lib/embed/media_tag.rb:
```diff
@@ -36,7 +36,7 @@ def to_html
 
     def media_element(label, file, type)
-      media_wrapper(label: label, stanford_only: file.stanford_only?, location_restricted: file.location_restricted?) do
+      media_wrapper(label: label, file: file) do
```
👍

jkeck commented on app/models/embed/purl.rb:
```diff
@@ -137,6 +138,35 @@ def initialize(msg = 'The requested PURL resource was not embeddable.')
+        unless errors.empty?
+          errors.each { |e| Rails.logger.warn e }
+          return nil
+        end
```
It feels like this method may have an extra responsibility that could be abstracted out.

What would you think about pulling this logic into private validation methods so we could do something along the lines of `return nil unless valid_duration?` at the top of `#to_s`?

Totally up to you.

jkeck commented:
Just one minor comment.  LGTM 👍

jmartin-sul commented:
ok, pushed up a new version with @jkeck's refactoring suggestion

ndushay commented:
LGTM
